### PR TITLE
[ELF] Respect orders of symbol assignments and DEFINED

### LINKED ISSUE #64600: [LLD] Evaluation of DEFINED linker script directive produces wrong output
The following linker script causes unexpected behaviour when `APPLE` is not defined by any input objects. The unexpected output is that `APPLE` will resolve to '1' instead of '2'.

```
APPLE = DEFINED (APPLE) ? 1 : 2;
```

It appears that when the symbol is first processed it is added to the symbol table with the expected value, however LLD reevaluates symbol assignments, and during these subsequent evaluations of the expression the `DEFINED` now returns 1 as `APPLE` has been inserted in the symbol table.

### PR DESCRIPTION
**Pull request by MaskRay**
Fix #64600: the currently implementation is minimal (see
https://reviews.llvm.org/D83758), and an assignment like
`__TEXT_REGION_ORIGIN__ = DEFINED(__TEXT_REGION_ORIGIN__) ? __TEXT_REGION_ORIGIN__ : 0;`
(used by avr-ld[1]) leads to a value of zero (default value in `declareSymbol`),
which is unexpected.

Assign orders to symbol assignments and references so that
for a script-defined symbol, the `DEFINED` results match users'
expectation. I am unclear about GNU ld's exact behavior, but this hopefully
matches its behavior in the majority of cases.

[1]: https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=ld/scripttempl/avr.sc

### CODE REVIEW

**Comment by benshi001** [comment]
LGTM. Would it be better to add a link to the GNU-ld's AVR default linker script, in your commit message ?

https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=ld/scripttempl/avr.sc;hb=HEAD

**Review by benshi001 — APPROVED** [review]

File: lld/ELF/LinkerScript.h [hunk 1]
```diff
@@ -86,9 +86,10 @@ struct SectionCommand {
 
 // This represents ". = <expr>" or "<symbol> = <expr>".
 struct SymbolAssignment : SectionCommand {
-  SymbolAssignment(StringRef name, Expr e, std::string loc)
+  SymbolAssignment(StringRef name, Expr e, std::string loc,
+                   unsigned symOrder = 0)
```
Comment (smithp35) [inline lld/ELF/LinkerScript.h:90]: There are only a couple more calls to the constructor in the same file. Would it be better to make the index explicit at the call site?

Not a strong opinion, and happy for it to be left as is.

File: lld/ELF/Config.h [hunk 2]
```diff
@@ -476,6 +476,12 @@ struct Ctx {
   // True if we need to reserve two .got entries for local-dynamic TLS model.
   std::atomic<bool> needsTlsLd{false};
 
+  // Each symbol assignment and reference is assigned an increasing order.
```
Comment (smithp35) [inline lld/ELF/Config.h:479]: By reference, do you mean the Defined(sym) reference. I couldn't see any other place that a symbol reference would get an increasing order.

If I'm right perhaps // Each symbol assignment and DEFINED(symbol) reference is assigned an increasing order.

**Review by smithp35 — APPROVED** [review]
I took a (very) quick look at GNU, it appears to have a statement counter that is used for a similar purpose.

Only got a couple of small comments so marking approved.

**Comment by MaskRay** [comment]
Thanks for the review. Improved comments.

**Pull request merged**